storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -367,102 +445,76 @@ where
     /// and this function will return `x`. Both `bits` and `siblings` start from the bottom.
     fn construct_subtree(
         bits: impl Iterator<Item = bool>,
-        siblings: impl Iterator<Item = Arc<SparseMerkleNode<V>>>,
-        leaf: Arc<SparseMerkleNode<V>>,
-    ) -> Arc<SparseMerkleNode<V>> {
+        siblings: impl Iterator<Item = SubTree<V>>,
+        leaf: SubTree<V>,
+    ) -> SubTree<V> {
         itertools::zip_eq(bits, siblings).fold(leaf, |previous_node, (bit, sibling)| {
-            Arc::new(if bit {
-                SparseMerkleNode::new_internal(sibling, previous_node)
+            if bit {
+                SubTree::new_internal(sibling, previous_node)
             } else {
-                SparseMerkleNode::new_internal(previous_node, sibling)
-            })
+                SubTree::new_internal(previous_node, sibling)
+            }
         })
     }
 
     /// Queries a `key` in this `SparseMerkleTree`.
     pub fn get(&self, key: HashValue) -> AccountStatus<V> {
-        let mut current_node = Arc::clone(&self.root);
+        let mut cur = self.root_weak();
         let mut bits = key.iter_bits();
 
         loop {
-            let next_node = if let Node::Internal(node) = &*current_node.read_lock() {
-                match bits.next() {
-                    Some(bit) => {
-                        if bit {
-                            node.clone_right_child()
-                        } else {
-                            node.clone_left_child()
+            if let Some(node) = cur.get_node_if_in_mem() {
+                if let Node::Internal(internal_node) = node.borrow() {
+                    match bits.next() {
+                        Some(bit) => {
+                            cur = if bit {
+                                internal_node.right.weak()
+                            } else {
+                                internal_node.left.weak()
+                            };
+                            continue;
                         }
+                        None => panic!("Tree is deeper than {} levels.", HashValue::LENGTH_IN_BITS),
                     }
-                    None => panic!("Tree is deeper than {} levels.", HashValue::LENGTH_IN_BITS),
                 }
-            } else {
-                break;
-            };
-            current_node = next_node;
+            }
+            break;
         }
 
-        let ret = match &*current_node.read_lock() {
-            Node::Leaf(node) => {
-                if key == node.key() {
-                    match node.value() {
-                        LeafValue::Value(value) => AccountStatus::ExistsInScratchPad(value.clone()),
-                        LeafValue::ValueHash(_) => AccountStatus::ExistsInDB,
+        let ret = match cur {
```

Comment:
maybe we can get rid of this

```
let ret = ...;
ret
```

structure now that things are written in a different way?